### PR TITLE
ECDSA Support

### LINKED ISSUE #50: Need updated cypto module and support for ECDSA keys
```
(pymanta) rui@hayek:~/git/python-manta$ ./bin/mantash -v ls /rui.loura/stor/
manta.client: DEBUG: ListDirectory '/rui.loura/stor'
manta.auth: DEBUG: sign 'date: Thu, 08 Nov 2018 23:15:21 GMT' with agent key (algo rsa-sha1, fp 5f:fb:db:f3:fd:43:ad:4e:ee:b6:24:75:74:a3:ae:8b)
manta.client: DEBUG: req: GET /rui.loura/stor
    host: us-east.manta.joyent.com
    headers: {'accept-encoding': 'gzip, deflate',
     'authorization': 'Signature keyId="/rui.loura/keys/5f:fb:db:f3:fd:43:ad:4e:ee:b6:24:75:74:a3:ae:8b",algorithm="rsa-sha1",signature="AAAAIQCVvdbep5IRuhlZRPPJxj9WOxbvI7iCwog09iwu8UmQcwAAACEAyjn0qbxCpTYjPhhkLM5XEVyvjUuBKPLfZkOvT3zlrIk="',
     'date': 'Thu, 08 Nov 2018 23:15:21 GMT',
     'user-agent': 'mantash/2.6.2 (linux2) Python/2.7.15rc1'}
    body: (none)
manta.client: DEBUG: res: GET /rui.loura/stor
    {'connection': 'keep-alive',
     'content-length': '99',
     'content-md5': 'o6kfsMz8ATeTfySea8Vqpg==',
     'content-type': 'application/json',
     'date': 'Thu, 08 Nov 2018 23:15:22 GMT',
     'server': 'Manta',
     'status': '403',
     'x-request-id': '2a723d10-e3ac-11e8-9e31-6357c012a6ae',
     'x-response-time': '6',
     'x-server-name': '39adec6c-bded-4a14-9d80-5a8bfc1121f9'}
    {"code":"InvalidSignature","message":"The signature we calculated does not match the one you sent"}
mantash: ERROR: (InvalidSignature) The signature we calculated does not match the one you sent
```

### PR DESCRIPTION
**Pull request by rjloura**
Fixes #50

### CODE REVIEW

**Comment by rjloura** [comment]
Beyond the tests I added, I also ran the test suite with `MANTA_KEY_ID`'s of:
* path to ecdsa-521 key
* removed password protected ecdsa-521 key from ssh-agent, re-ran test suite.
* SHA256 fingerprint of RSA key with leading `SHA256:`
* MD5 fingerprint of ECDSA key with leading `MD5:`
* MD5 fingerprint of ECDSA key without leading `MD5:`

File: manta/auth.py [hunk 1]
```diff
@@ -183,21 +218,27 @@ def load_ssh_key(key_id, skip_priv_key=False):
         pub_key_path=pub_key_path,
         fingerprint=fingerprint,
         priv_key_path=priv_key_path,
-        priv_key=priv_key)
+        priv_key=priv_key,
+        algorithm=algo)
+
+def rsa_sig_from_agent_sign_response(response):
```
Comment (trentm) [inline manta/auth.py]: Consistency nits: "...from_agent_sign_response" here, but from "...*signed*..." response in the next function. 
Also:
- `response` v `d`
- `algo` v `alg`

File: requirements.txt [hunk 2]
```diff
@@ -1,3 +1,3 @@
-pycrypto == 2.6
+cryptography >= 2.3.1
```
Comment (trentm) [inline requirements.txt]: Should the SmartOS installation notes about installing the "py27-crypto" pkgsrc package be updated?

File: README.md [hunk 3]
```diff
@@ -309,6 +240,30 @@ For other limitations (also planned work) see TODO.txt.
 
 An attempt to cover some common install/setup issues.
 
+## `pynacl` dependency install failure
```
[the remark this answers is not in the document]
Comment (rjloura) [reply README.md]: done

File: manta/auth.py [hunk 4]
```diff
@@ -392,11 +462,18 @@ def sign(self, s):
 
         key_info = self._get_key_info()
         assert key_info["type"] == "agent"
+
+        # XXX: Do we really need this check?
```
[the remark this answers is not in the document]
Comment (rjloura) [reply manta/auth.py]: The virtualenv where I did most of the testing for this has paramiko 2.4.2.  The same version is installed on my SmartOS zone, which I tested as well.  I bumped the version in requiements.txt to `>=2.4`, and removed the checks for PARAMIKO_VER_INFO.

File: manta/auth.py [hunk 5]
```diff
@@ -131,16 +150,25 @@ def load_ssh_key(key_id, skip_priv_key=False):
...
+        except:
+            raise MantaError("Unsupported key type for: {}".format(key_id))
+
         return dict(
             pub_key_path=pub_key_path,
             fingerprint=fingerprint,
             priv_key_path=key_id,
-            priv_key=priv_key)
+            priv_key=priv_key,
+            algorithm=algo)
```
[the remark this answers is not in the document]
Comment (rjloura) [reply manta/auth.py]: done

File: manta/auth.py [hunk 6]
```diff
@@ -131,16 +150,25 @@ def load_ssh_key(key_id, skip_priv_key=False):
         finally:
             f.close()
         fingerprint = fingerprint_from_ssh_pub_key(pub_key)
+
+        # XXX: pubkey should NOT be in PEM format.
+        try:
+            algo = SSH_KEY_MAP[pub_key.split()[0]]
+        except:
```
[the remark this answers is not in the document]
Comment (rjloura) [reply manta/auth.py]: done in both places.

File: manta/auth.py [hunk 7]
```diff
@@ -37,8 +33,29 @@
...
+
+# TODO: add and test other ECC curves
+SSH_KEY_MAP = {
+    "ecdsa-sha2-nistp256": ECDSA_SHA256_STR,
+    "ecdsa-sha2-nistp384": ECDSA_SHA384_STR,
+    "ecdsa-sha2-nistp521": ECDSA_SHA512_STR,
+    "ssh-rsa": RSA_STR
+}
 
+ECDSA_HASH_MAP = {
```
[the remark this answers is not in the document]
Comment (rjloura) [reply manta/auth.py]: That was well worth the 10 min read.

File: manta/auth.py (hunk 1 above)
Comment (rjloura) [reply manta/auth.py]: done.

File: requirements.txt (hunk 2 above)
Comment (rjloura) [reply requirements.txt]: I replaced that note with a note about installing libsodium and pynacl.

**Comment by trentm** [comment]
https://pypi.org/project/manta/2.7.0/ is now published with this.

**Pull request merged**